### DISCUSSION TITLE
Pause read-sides on application start

### DISCUSSION THREAD

**Discussion by apeshimam** [chief-of-state/chief-of-state#32] [Ideas]
# What is the need
We would like to be able to add a new read handler without it replaying all of the events starting from the beginning.

# Research

There are two tables related to read sides that are created: `read_sides` and `read_side_offsets`.  `read_sides` actually stores the state indicating if a read side is paused. `read_side_offsets` store the current offset for each read handler + shard.  It is important to note that these tables are populated lazily. 

I have successfully paused the read handler in `ServiceBootstrap` by doing the following (obviously for live scenarios we would read the projection name from config):
```
    readSideStateManager.pauseForAll("READSIDE_1").map(result => log.info(f"pause result: $result")) 
```
The effect of this is to insert into `read_sides` the paused state for each projection name and key. Here's the state from my local db.

```
postgres=# select * from read_sides;
 projection_name | projection_key | paused | last_updated
-----------------+----------------+--------+---------------
 READSIDE_1      | 7              | t      | 1650481603029
 READSIDE_1      | 5              | t      | 1650481603028
 READSIDE_1      | 2              | t      | 1650481603029
 READSIDE_1      | 0              | t      | 1650481603029
 READSIDE_1      | 3              | t      | 1650481603042
 READSIDE_1      | 6              | t      | 1650481603029
 READSIDE_1      | 4              | t      | 1650481603031
 READSIDE_1      | 8              | t      | 1650481603039
 READSIDE_1      | 1              | t      | 1650481603033
```

Additionally, we should note that the `read_side_offsets` table remains empty:
```
postgres=# select * from read_side_offsets;
 projection_name | projection_key | current_offset | manifest | mergeable | last_updated
-----------------+----------------+----------------+----------+-----------+--------------
(0 rows)
```

What we can do is to inspect the config before enabling the read_sides and pause any read handlers that should be paused. 

Our goal remains to update the offset of the read_side, which we can do by utilizing the ProjectionManagement API. Note that will require adding a function to pass an in integer representing the current offset. Currently, only `skipOffset` is available. 

Once this is complete, we can resume operation of the Read Handler.

**Comment by Tochemey** [comment]
Yeah it will be straightforward to pause a read side or all read sides on app start. Offsets are only available when read sides consume events. Since that is not the case we might have to resort to a custom offset computation and play with that. I think @zenyui has some thoughts around that. Also in that regards we need to take into consideration the actual projection offset computation not to mess up the read side when they start consuming events.

**Comment by Tochemey** [comment]
@apeshimam I will suggest we get this requirement in two phases:

1. Start all read-sides or some read-side in paused state(low hanging fruit)
2. Start read-side at a given offset(a bit involving)
